### PR TITLE
Injectable renderer

### PR DESCRIPTION
**Pull request by justsml**
Supports different input configuration, and more granular theming.

### CODE REVIEW

File: src/react-sortable-tree.js [hunk 1]
```diff
@@ -361,7 +363,7 @@ class ReactSortableTree extends Component {
     }
 
     renderRow({ node, path, lowerSiblingCounts, treeIndex }, listIndex, key, style, getPrevRow, matchKeys) {
-        const NodeContentRenderer = this.nodeContentRenderer;
+        const NodeContentRenderer = this.state.nodeContentRenderer || this.props.nodeContentRenderer;
```
Comment (fritz-c) [inline src/react-sortable-tree.js]: When you use `this.props.nodeContentRenderer` directly, it won't have the props related to drag-and-drop injected into it. It needs to be wrapped with `dndWrapSource`, as in the constructor.

I made a decision when writing this that the `nodeContentRenderer` would only be referenced once, at component mount time. As it won't respond to changes, I kept it out of `this.state`.

File: src/content-renderer.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+
+export const contentRenderer = function({node, styles, path, treeIndex, buttons}) {
```
Comment (fritz-c) [inline src/content-renderer.js]: As `contentRenderer` is quite close to `nodeContentRenderer`, I'd prefer you change the name to `innerContentRenderer`.

Also, do you mind making this `export default ...` just to be more consistent with other files?

One last naming thing - please change the file name to `default-inner-content-renderer.js` to better represent its significance.

File: src/react-sortable-tree.js [hunk 3]
```diff
@@ -471,6 +474,7 @@ ReactSortableTree.propTypes = {
     // This is an advanced option for complete customization of the appearance.
     // It is best to copy the component in `node-renderer-default.js` to use as a base, and customize as needed.
     nodeContentRenderer: PropTypes.any,
+    contentRenderer: PropTypes.func.isRequired,
```
Comment (fritz-c) [inline src/react-sortable-tree.js]: Could you add a comment explaining the purpose here? The same explanation should go in the README.md as well.

File: src/react-sortable-tree.js (hunk 1 above)
Comment (justsml) [inline src/react-sortable-tree.js]: Thx for that explanation, I'm new to this React stuff still :)

File: src/content-renderer.js (hunk 2 above)
Comment (justsml) [inline src/content-renderer.js]: Np

File: src/react-sortable-tree.js (hunk 3 above)
Comment (justsml) [inline src/react-sortable-tree.js]: Should this be `innerContentRenderer`?

Comment (justsml) [inline src/react-sortable-tree.js]: Progressive overrides would help me with some design ideas I have for my JSON editor.

In particular, I'd like to 3d-flip the tree node in-place to reveal an edit box. Right now I'm about to implement it into a modal - I'm just not liking my progress so far... 💩

Comment (fritz-c) [inline src/react-sortable-tree.js]: Yeah, this should be `innerContentRenderer`.

**Comment by fritz-c** [comment]
@justsml 
Thanks for the filename change. Can you make the export statement use `export default ..` in that file, just for consistency?

It looks like you deleted `innerContentRenderer` from the propTypes of the `ReactSortableTree` component. I thought this would be added at the top so users can customize it. If so, it should be in the propTypes there. Also, a prop name of `innerContentRenderer` would be preferable to the `contentRenderer` you are currently using in that file (again, just for consistency).

**Comment by fritz-c** [comment]
I just thought of something - how about instead of handing the `innerContentRenderer` as a prop to the `NodeContentRenderer`, you make it the child of the node?

e.g.,
```jsx
const InnerContentRenderer = this.props.innerContentRenderer;

return (
    <TreeNode
        style={style}
        // ...
        dragHover={this.dragHover}
    >
        <NodeContentRenderer
            node={node}
            // ...
            {...nodeProps}
        >
            <InnerContentRenderer
                node={node}
                // ...
            />
        </NodeContentRenderer>
    </TreeNode>
);
```

One issue is that you won't be able to hand over `styles` as easily, but I feel like if you're separating the inner content from the rest of the node, it should have its own stylesheet.

**Comment by justsml** [comment]
Hey @fritz-c thx for the review.

I pushed some updates - but I missed your implementation example of `<InnerContentRenderer />` 
I'll try refactor it in later today.

**Comment by fritz-c** [comment]
Going to close this due to inactivity.

**Pull request closed without merging**